rfcs/2024-08-21-motion-presets.md

```diff
@@ -0,0 +1,555 @@
+};
+```
+
+</details>
+
+### Highlight Animations
+
+#### Morph
+
+> [!Note]
+>
+> Defining morph as preset is a bit tricky and need to make sure the API is feasible with all possible scenarios. Currently we've done a basic feasibility check but we might change / update the API if we come across some animation that can't be built with this API.
+
+<table>
+<tr>
+  <td width="600px">
+
+<!-- prettier-ignore -->
+```jsx
+import { AnimatePresence } from 'framer-motion';
+import { 
+  Button, 
+  Morph 
+} from '@razorpay/blade/components';
+
+<AnimatePresence>
+{
+  isChatVisible ? (
+    <Morph layoutId="chat-interface">
+      <ChatInterface />
+    </Morph>
+  ) : (
+    <Morph layoutId="chat-interface">
+      <Button icon={RazorpayIcon} />
+    </Morph>
+  )
+}
+</AnimatePresence>
+```
+
+  </td>
+  <td width="500px">
+
+https://github.com/user-attachments/assets/2ec967a5-1e17-443a-a3a9-230962f6bd9a
+
+_Previews are just examples of presets. They don't use actual durations and easings yet_
+
+  </td>
+</tr>
+</table>
+
+<details>
+<summary>Alternate Morph APIs</summary>
+
+#### 2. Using `motion` from framer-motion
+
+```jsx
+import { motion } from 'framer-motion';
+import { Heading } from '@razorpay/blade/components';
+
+const CardHeading = motion(Heading);
+
+<CardHeading layoutId="card-heading" transition={{ duration: theme.motion.duration.slow }}>Hello, World!</CardHeading>
+<CardHeading as="h1" layoutId="card-heading">Hello, World!</CardHeading>
+```
+
+**Cons:**
+
+- We won't be able to preset the animation styles and durations with this approach
+- Can lead to inconsistent animations
+- Inconsistent with other motion presets so not very intuitive and requires learning framer-motion for syntax
+
+#### 3. Exposing `morph` Function from Blade
+
+```jsx
+import { Heading, morph } from '@razorpay/blade/components';
+
+const CardHeading = morph(Heading);
+
+<CardHeading layoutId="card-heading">Hello, World!</CardHeading>
+<Card>
+  <CardBody>
+    <CardHeading as="h1" layoutId="card-heading">Hello, World!</CardHeading>
+    <Box>
+      <Text>Other Text</Text>
+    </Box>
+  </CardBody>
+</Card>
+```
+
+**Pros:**
+
+- Allows presetting animation properties on blade
+
+**Cons:**
+
+- Inconsistent with other motion presets so not very intuitive
+- Comparitively more verbose than suggested API
+
+</details>
+
+#### AnimateInteractions
+
+When we wrap a certain component in AnimateInteractions wrapper from blade, we can animate the children component on interactions of the parent component.
+
+#### Scale
+
+Scale animation can be used indepedently to scale item on certain actions but also inside AnimateInteractions.
+
+<table>
+<tr>
+  <td width="700px">
+In below example, the images scales up when its parent container is hovered<br/><br/>
+
+<!-- prettier-ignore -->
+```jsx
+import { 
+  AnimateInteractions, 
+  Scale 
+} from '@razorpay/blade/components';
+
+<AnimateInteractions motionTriggers={['hover']}>
+  <Card>
+    <CardBody>
```

Comment:
Will AnimateInteraction create a new div wrapper? or is this a enhancer component?

Reply:
I had to create a div wrapper because something was not working. Lemme check again and see if it can be enhancer component